Sdtrig.tex

```diff
@@ -319,6 +319,37 @@ \subsection{Cache Operations}
     \end{steps}
 \end{commentary}
 
+\subsection{Invalid Virtual Addresses}
+
+For address matches without a mask, \RcsrTdataTwo must be able to hold all valid
+addresses in all supported translation modes. That means that after writing any
+of these valid addresses, the exact same value is read back, including any high
+bits.
+
+\begin{commentary}
+    If physical addresses are less than XLEN bits wide, they are zero-extended.
+    If virtual addresses are less than XLEN bits wide, they are sign-extended.
+    Be sure to have enough flops in \RcsrTdataTwo to hold all XLEN bits of the
+    full range of both physical and virtual addresses.
+\end{commentary}
+
+If \RcsrTdataTwo can hold any invalid addresses, then writes of an
+invalid address that can not be represented as-is should be converted to
+a different invalid address that can be represented.
+
+For invalid instruction fetch and load and store effective addresses,
+the compare value may be changed to a different invalid address.
```

Comment:
This is confusing.  It appears to say that an ifetch address or ldst effective address, if invalid, can be changed to a different invalid address value to create the compare value that is compared against te trigger address.  I assume that is not the arch intent trying to be expressed here.

Or if that is the intent, then who would actually choose to remap invalid compare addresses to different invalid values before doing a compare (i.e. no one would).

It seems like all reasonable implementations would simply compare the XLEN-wide ifetch/ldst address with the  XLEN-wide trigger address.  Is there some other implementation that the spec is trying to allow for that would justify being rather vague about exactly hwo copares are performed.

Reply:
We discussed a case in the last meeting where, in Sv48, there is a jump to an invalid address, say 0x1234aaaabbbbcccc.  That target address may be converted to another invalid address, presumably one that can be held in *tval (likely 49-bit 0x0aaaabbbbcccc), before it is fetched and compared with triggers.  Are you saying you think this should be a non-compliant implementation?

Reply:
Restating @bcstrongx's example:
1. Jump to an invalid address 0x1234aaaabbbbcccc.
2. It seems likely the PC cannot represent all possible invalid addresses, so the invalid address is converted to some internal representation.
3. So the instruction fetch must happen at some invalid address, but likely not the one generated by the original instruction.

Since that is allowed, it's reasonable to allow the same for load/stores as well.

Reply:
I believe I get the point.  That any "remapping" of an invalid compare address to another invalid address is more a matter of canonicity decode logic recognizing that the upper bits are non-canonical and one would, for example, just carry forward through the datapath one bit to signify valid vs. invalid address.  Then, when it gets to comparing this with an XLEN-wide tdata2 address value, the implementation could turn the one bit back into some fixed invalid pattern of high bits.

If that's the kind of idea you all have in mind, then that makes sense.  And hence the idea of invalid ifetch/ldst addresses being remappable to other invalid addresses before being compared with tdata2, seems valid and something to allow for.

Although it would seem like a "good" and cheap implementation would remap invalid address values written into tdata2 into an internal representation with an extra bit signifying valid vs. invalid address.  And invalid ifetch/ldst address values would be represented, as mentioned above, with one extra bit.  Then comparisons would indicate a match if either both addresses are valid and fully match all supported address bits, or both addresses are invalid and are then considered to match irrespective of any comparison of the lower supported address bits.  (With this, the address comparator is not XLEN-wide.  Versus converting address values to XLEN-wide values and then comparing them.  Although one can say that these are functionality equivalent implementations if both conversions of invalid addresses to XLEN-wide values do the same expansion.)

This provides a simple easy to understand model to software as to trigger matching behavior wrt invalid addresses.  Personally I would maybe mandate this behavior, but I expect others want a much looser spec that allows other behaviors that an even cheaper (?) implementation might exhibit.  (And, to be clear, an implementation would be allowed to instead never match whenever either or both addresses is/are invalid.)

Reply:
I'd be OK saying an implementation should map invalid addresses to a canonical invalid representation, that can also be represented in tdata2. But they're also allowed to have more flexible less useful behavior (for backwards compatibility) or may not match against invalid addresses at all.

Thoughts?

Reply:
I feel we're getting off-track trying non-normatively specifying desired invalid address encodings.

Let's focus on the normative part of the spec. The language says:
> For invalid instruction fetch addresses and load and store effective addresses,
> the compare value may be changed to a different invalid address.

That's clear to me but it was not clear to @gfavor, although he now understand what we're trying to say. How should this language be changed so its meaning is clear?

Reply:
> I'd be OK saying an implementation should map invalid addresses to a canonical invalid representation, that can also be represented in tdata2. But they're also allowed to have more flexible less useful behavior (for backwards compatibility) or may not match against invalid addresses at all.

Validity isn't known a priori.  If you support Sv48 and 48b PA (the example from the emails) then 0x0000FFFFFFFFFFFF is a valid bare mode address but not a valid Sv48 address.  So you don't really know if it's valid until the time of use.  The trigger might be enabled in all modes (M/S/U/VS/VU) so it's valid sometimes (M) and invalid other times (U if satp.MODE=Sv48).  Or satp.MODE can change dynamically which changes the way it's interpreted.

Reply:
It looks fine to me.  Do we want to say anything (non-normatively) about how, ideally, the conversion of invalid addresses for both tdata2 and the address to be compared are the same, though it is not guaranteed?

Reply:
I suggest taking the one extra day to properly resolve this issue and not temporarily sweeping it under the rug.  You're not actually saving time to get through the steps to Freeze approval.  It will still be a change that will need to be reviewed by AR (and in the process you wil need to specifically inform the ARC of changes and probably provide a delta spec).  So please just save everyone's time.

Reply:
I'm not taking a position on this; the rest of the TG should collectively decide.  But I'll suggest that Beeman or someone else, if they desire, create a PR for a non-normative comment.  That will concretely and expeditiously move the ball forward.  Rushing to finish things should not be the primary reason for not doing something.  Do what's right for the community and all the readers and implementers of the Debug spec that are going to come along next year and the many following years.

Reply:
If that's going to be a separate PR then is this one ready to wrap up?